modules/compliance-tests/src/smithy4s/compliancetests/internals/package.scala

```diff
@@ -29,4 +33,33 @@ package object internals {
       h.++(Hints(smithy.api.TimestampFormat.EPOCH_SECONDS.widen))
     )
   }
+  
+  def splitQuery(queryString: String): (String, String) = {
+    queryString.split("=", 2) match {
+      case Array(k, v) =>
+        (
+          k,
+          Uri.decode(
+            toDecode = v,
+            charset = StandardCharsets.UTF_8,
+            plusIsSpace = true
+          )
+        )
+      case Array(k) => (k, "")
+    }
+  }
+
+  def parseQueryParams(
+      queryParams: Option[List[String]]
+  ): ListMap[String, List[String]] = {
+    queryParams.combineAll
+      .map(splitQuery)
+      .foldRight[ListMap[String, List[String]]](ListMap.empty) {
```

Comment:
this seems awfully close to `foldMap` using the map monoid (but it won't keep the ordering like ListMap :( )

Reply:
TIL about ListMap

Reply:
I need to preserve order